### ISSUE TITLE
Regarding training on segments

### ISSUE THREAD

**Issue by ayushjain1144** [JonasSchult/Mask3D#56]
Hi Jonas,

A follow up to the above question: how critical it is for performance (beyond memory consumption) to train/evaluating on segments instead of randomly sampling points? (Just like image mask2former)
And what if instead of segments, one just "voxelize" the point cloud and keep a single point from the whole voxel?

Any insights would be really helpful!

Thanks
Ayush

**Comment by JonasSchult** [comment]
Hi Ayush,

thanks for your interest in our project! It was nice chatting with you at ECCV!

Performance stays roughly the same when training on segments. Memory consumption decreases, though.

> And what if instead of segments, one just "voxelize" the point cloud and keep a single point from the whole voxel?

Our `Sampled Cross-Attention` (p. 4 in the [paper](https://arxiv.org/pdf/2210.03105.pdf) ) operates on voxels and not individual points. Going back to points would increase memory consumption drastically. (Some approaches such as SoftGroup++ call it "Late Devoxelization").

Hope I could help! Let me know if you have questions. :)

Best,
Jonas

**Comment by ayushjain1144** [comment]
Hi Jonas!

Thank you for your very helpful response, and it was great meeting you as well!!

My understanding of segments is that its only used during dot product of queries and voxel features. Did you also try randomly sampling voxel points to do this? something similar to what mask2former does for backproping loss only at a subset of points? mask2former uses logits for finding uncertain points which requires them to do a dot product on all points first and hence its not really applicable here, but do you know how would pure random sampling of points to do dot product and compute  loss would do?

**Comment by JonasSchult** [comment]
Hi!

> Did you also try randomly sampling voxel points to do this? something similar to what mask2former does for backproping loss only at a subset of points?

This is a super interesting direction! Mask2Former relies on [PointRend](https://arxiv.org/abs/1912.08193) to reduce the memory consumption.
A naive implementation of PointRend for 3D point clouds would require to tri-linearly interpolate point features. ([Mask2Former code](https://github.com/facebookresearch/detectron2/blob/main/projects/PointRend/point_rend/point_features.py#L39)). Tri-linear interpolation is already implemented in the MinkowskiEngine, however, I assume that the sparsity of the point cloud can cause problems here (free space is dominating). I didnt conduct any experiments in this direction, but I think that a more 3D-specific flavour of PointRend could be interesting here.

> but do you know how would pure random sampling of points to do dot product and compute loss would do?

I played a bit around with pure random sampling. However, I couldn't get better (or more memory-friendly) results with this. However, I did not invest a lot of time in this direction. So, there is a chance that with proper tuning better results are possible. :)
As a starting point, you can change the number of randomly sampled points [here](https://github.com/JonasSchult/Mask3D/blob/main/conf/matcher/hungarian_matcher.yaml#L6).


Best,
Jonas

**Comment by ayushjain1144** [comment]
Thank you so much! This was very helpful!
